### PR TITLE
refactor(coap-server): refactor request handling

### PR DESCRIPTION
**Pull request by JKRhb**
This PR applies some additional refactoring to the `CoapServer` class, removing some nesting from the `handleRequest` method. In follow-up PRs, the logic of this and other methods needs to be further refactored in order to reduce complexity and make the class more maintainable.

### CODE REVIEW

**Comment by JKRhb** [comment]
I think this PR should now finally be ready for review :) Unfortunately, the PR size turned out to be a lot larger than originally intended. However, I think the code should now be significantly more concise and hopefully also more readable, which should make it easier to extend the CoapServer in the future.

Some helper functions might be able to be generalized and moved to the core package in follow-up PRs. Also, there are still some workarounds in the code referring to bugs or missing features in node-coap, some of which might be fixed in the meantime. I will investigate that and will otherwise –  if the issues should still be present – try to resolve them.

File: packages/binding-coap/src/coap-server.ts
Comment (relu91) [inline packages/binding-coap/src/coap-server.ts:42]: Why doesn't `WoT.InteractionOptions` have `formIndex` already? Current scripting API already mention it: https://w3c.github.io/wot-scripting-api/#dom-interactionoptions-formindex also in TS file they are indeed present:https://github.com/w3c/wot-scripting-api/blob/06a31b1efb06d8b1ec996cd18dab25c71a98b2b1/typescript/scripting-api/index.d.ts#L175.

File: packages/binding-coap/src/coap-server.ts
Comment (relu91) [inline packages/binding-coap/src/coap-server.ts:521]: Sorry for the current formatting is hard to tell, but can't we use `trycatch` and `await` here?

**Review by relu91 — APPROVED** [review]
I have only questions but the content of the PR looks good!

File: packages/binding-coap/src/coap-server.ts
Comment (JKRhb) [reply packages/binding-coap/src/coap-server.ts]: Good point, that has been a part of the original implementation which I found kind of strange therefore I didn't want to touch it initially 😅 I now tried to provide a proper solution as you proposed in 3b8604c4eb20954802c338b363e964f40956796e. Not sure if always invoking `res.end()` as in the original code is the correct approach here, though.

Comment (JKRhb) [reply packages/binding-coap/src/coap-server.ts]: Slightly adjusted the code in 0bf0d5af70fbfbefcac2bc238084574a74b7cf6a to be closer to the original logic.

File: packages/binding-coap/src/coap-server.ts
Comment (danielpeintner) [reply packages/binding-coap/src/coap-server.ts:42]: Agree, let look into it in a follow-up. Created #1039

File: packages/binding-coap/src/coap-server.ts
Comment (danielpeintner) [inline packages/binding-coap/src/coap-server.ts:354]:
```suggestion
        debug(`Request contained accept option with value ${accept} which is supported (or default/empty).`);
```

I struggled over this debug statement and it confused me. I tried to explain it better but I have difficulties to convey the message 🤷‍♂️

File: packages/binding-coap/src/coap-server.ts
Comment (danielpeintner) [inline packages/binding-coap/src/coap-server.ts:512]:
```suggestion
                error(`CoapServer on port ${this.getPort()} got internal error on observe '${req.url}': ${err.message}`);
```

Shouldn't it be observe?

File: packages/binding-coap/src/coap-server.ts
Comment (danielpeintner) [inline packages/binding-coap/src/coap-server.ts:647]: I think the debug statement might convey that it deals with an event

File: packages/binding-coap/src/coap-server.ts
Comment (danielpeintner) [inline packages/binding-coap/src/coap-server.ts:775]: VERY MINOR
In the following code there are some parts that always read like

```
        res.code = ...;
        res.end(payload);
```

It could make sense to have an overarching method that deals with all of them.. so that all point to it..

**Review by danielpeintner — APPROVED** [review]
PR LGTM

Some comments below.. really just comments

File: packages/binding-coap/src/coap-server.ts
Comment (relu91) [reply packages/binding-coap/src/coap-server.ts:42]: Thank you daniel, we can resolve this issue.

File: packages/binding-coap/src/coap-server.ts
Comment (JKRhb) [reply packages/binding-coap/src/coap-server.ts]: Good point, I now turned the `sendErrorResponse` method into a generic `sendResponse` method :)

File: packages/binding-coap/src/coap-server.ts
Comment (JKRhb) [reply packages/binding-coap/src/coap-server.ts:663]: Would this be a better alternative?

```suggestion
                `CoapServer on port ${this.getPort()} rejects '${affordanceKey}' event subscription from ${Helpers.toUriLiteral(
```

Comment (JKRhb) [reply packages/binding-coap/src/coap-server.ts]: Took the freedom of already adding this suggestion to the PR as part of 5d0f086c474935df016f5c0b5031e1798ace0328, hope that was okay :)

**Review by relu91 — APPROVED** [review]
Reviewed once again, looks good!

**Comment by JKRhb** [comment]
> Reviewed once again, looks good!

Awesome, thank you :) I just squashed the commits into one – the PR would be ready to merge when you are :) CC @danielpeintner

**Comment by JKRhb** [comment]
> So in general, we had this non-written rule that once two committers agree on a PR we can merge it, but given the fact that daniel explicitly comment on this PR I would wait. It is not fixing any critical bug and there is no urgency. Is that ok @JKRhb?

Sure, from my side we can wait for Daniel's final approval :)

**Review by danielpeintner — APPROVED** [review]

**Comment by danielpeintner** [comment]
PR looks good, thanks! I will merge it right away ...

> Awesome, thank you :) I just squashed the commits into one – the PR would be ready to merge when you are :) CC @danielpeintner

Side comment: I think we should avoid squashing commits unless there is a good reason.
Why do I say that? Squashing commits does no longer allow reviewers to check what has been updated recently. Hence, one needs to re-check everything again...
Having said that, I like it to squash the commits when merging the PR because doing so

1. Reviewer still see the history
2. The commit in master/main is still clean and just contains *one* commit even if the PR evolved a lot.

What do people think. Shall we setup some kind of rules for that?

**Comment by danielpeintner** [comment]
What I would like to propose is the following.

* while we work on a PR and people are reviewing we should **not**  squash commits
* once we agree to merge the PR we can (and maybe should) squash all the commits into one. Hence we squash in the very last step.
 
image

**Comment by relu91** [comment]
I'm ok with that, but only if the changes are closely related. I would not squash if the PR is fixing two issues or dealing with multiple aspects of the code base (e.g. refactoring + fixing).

**Comment by JKRhb** [comment]
You are right @danielpeintner, sorry for squashing a bit prematurely :/ Next time, I will wait until all final approvals are given.

I guess as a general rule, maybe we can say that squashing should only be done when the reviewers gave their approval/ask a PR author to do so or during the merge process (as you've shown above).

**Pull request merged**
